src/chains/ethereum/console.log/src/handlers.ts

```diff
@@ -0,0 +1,152 @@
+import { bufferToBigInt } from "@ganache/utils";
+
+// bytes1 - bytes32
+type Digits = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;
+type Numbers = Digits | 10 | 20 | `${1}${Digits}` | `2${Digits}` | 30 | 31 | 32;
+export type BytesN = `bytes${Numbers}`;
+export type FixedBytesN = `fixedBytes(${Numbers})`;
+export type Handler = keyof typeof Handlers | FixedBytesN;
+
+/**
+ * EVM word size for abi encoded data
+ */
+export const WORD_SIZE = 32 as const; // bytes
+
+/**
+ * Used to convert _from_ twos compliment
+ * This is calculated via:
+ * ~((1n << (8n * BigInt(WORD_SIZE))) - 1n)
+ */
+const OR_WITH_TWOS_COMPLEMENT =
+  -0x10000000000000000000000000000000000000000000000000000000000000000n as const;
+
+/**
+ * For dynamic-length types, like `bytes` and `string`, returns the starting
+ * position of the value in `memory`, and the `end` position of the value.
+ *
+ * Calling this function when the type is not dynamic will not work. You have
+ * been warned.
+ *
+ * @param memory
+ * @param offset The offset into the given memory from which to start parsing.
+ * @returns The starting position of the value in `memory`, and the `end` position of the value.
+ */
+function getDynamicDataMarkers(memory: Buffer, offset: number) {
+  // A Buffer's length maxes out at UInt32 max, 4294967295, (2**(8 * 4)) - 1) it
+  // is safe to decode the `start` and `length` values as UInt32s.
+  //
+  // If we had a start position "word" of:
+  //   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 60
+  // we know we really only need to read the last 4 bytes, since we know that
+  // a Buffer's length can't be greater than UInt32 (which can be represented by
+  // up to 4 bytes). So we only need to look at:
+  // 00 00 00 60
+  // `offset + WORD_SIZE - 4` effectively tells `readUInt32BE` to read only
+  // these last 4 bytes of the word that represent the `start`
+  // value.
+  // We add WORD_SIZE to this value because without it it doesn't work.
```

Comment:
If I'm understanding correctly, `memory.readUInt32BE(offset + WORD_SIZE - 4)` gets us the last 4 bytes (and only relevant bytes) of the word after the offset, which contains the location of the data we actually care about here. At that location is a single word containing the length of the data, plus our actual data.

By adding ` + WORD_SIZE + 4`, we're skipping the word that represents the length (and, as your comment says, the signature) so we actually start at the data. Conveniently, we're only looking for the last 4 bytes of the "length" word, so we can just subtract 4 to get the length of the data. Neat stuff!

If this is correct, can we update the 
> We add WORD_SIZE to this value because without it it doesn't work.

comment to indicate that we are skipping the length of the data to get to the actual start?

Also, `4` is a magic number here that comes up a lot, but it's more a coincidence that it's 4 in different cases. For example, we add 4 because this is the size of the signature. We subtract 4 because we move to the _end_ of the word, then backtrack to the relevant data. Could we perhaps be more explicit about what is happening and where data lives?

```TypeScript
  const SIGNATURE = 4;
  const SHORTCUT_TO_LAST4 = WORD_SIZE - 4;
  const startOfLength = memory.readUInt32BE(offset + SHORTCUT_TO_LAST4) + SIGNATURE;
  const start = startOfLength + WORD_SIZE; // the data starts after the length, which is one word
  const length = memory.readUInt32BE(startOfLength + SHORTCUT_TO_LAST4);
  return { start, end: start + length };
```

or a smaller change that only gives the signature a name:
```TypeScript
  const SIGNATURE = 4;
  const start = memory.readUInt32BE(offset + WORD_SIZE - 4) + WORD_SIZE + SIGNATURE;

  // the same optimization is applied here.
  // The word that immediately precedes the data itself is the data's `length`:
  const length = memory.readUInt32BE(start - 4);
  return { start, end: start + length };
```

Reply:
Good call. Simplified!